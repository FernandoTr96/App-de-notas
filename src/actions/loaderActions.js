import { types } from "../types/types"

export const startLoader = ()=>{
    return {
        type: types.loaderTrue,
        payload:{
            value: true
        }
    }
}

export const finishLoader = ()=>{
    return {
        type: types.loaderFalse,
        payload:{
            value: false
        }
    }
}

export const purgeLoader = ()=>(
    {
        type: types.purgeLoader
    }
)