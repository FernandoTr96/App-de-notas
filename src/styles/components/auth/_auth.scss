@import '../../_classes/buttons.scss';
@import '../../_classes/main_content.scss';
@import '../../_mixins/mq_breakpoints.scss';

.auth__main{
    @extend %main-content;
    overflow-x: hidden;
    background-image: url('https://cdn.pixabay.com/photo/2013/10/07/18/06/field-192179_960_720.jpg');
    background-repeat: no-repeat;
    background-size: cover;
    justify-content: center;
    align-items: center;
}

.auth__form-container{
    width: 100%;
    height: 100vh;
    padding: 1.5rem;
    box-shadow: 3px 3px 10px 0px $shadow;
    border-radius: .3rem;
    background-color: #fff;

    @include mobile{
        width: 70%;
        height: auto;
    }

    @include tablet{
        width: 50%;
        height: auto;
    }
    @include desktop{
        width: 40%;
        height: auto;
    }
    @include large{
        width: 30%;
        height: auto;
    }
}

.auth__form{
    
    width: 100%;
    display: flex;
    flex-direction: column;

    legend{
        font-weight: bold;
        font-size: 1.5rem;
        color: darken($light, 55);
        padding: .5rem;
        text-align: center;
    }

    input{
        display: block;
        padding: .5rem;
        margin: 1rem 0rem;
        color: darken($light,40);
        font-size: medium;
        font-family: serif;
        border: none;
    
        &:focus{
            outline: none;
        }
    
        &:invalid{
            border-bottom: 1px solid $primary;
        }
    
        &:valid{
            border-bottom: 1px solid green;
        }
    }

    button[type='submit']{
        @extend %buttonBlock;
        &:hover{
            transition: background 500ms;
            background-color: darken($primary, 5);
        }
    }
}

.auth__google-legend{
    font-size: small;
    text-align: center;
    margin: .5rem auto;
    color: $primary;
    font-weight: 600;
}
.auth__google-button{
   
    @extend %buttonSocialNetwork;

    &:hover{
        transition: background 500ms;
        background-color: darken($light,3);
    }
    img{
        width: 1.5rem;
        height: auto;
        object-fit: cover;
        margin-right: .5rem;
    }
}
.auth__link{
    margin: 1rem auto;
    font-size: small;
    color: darken($primary,10);
}
