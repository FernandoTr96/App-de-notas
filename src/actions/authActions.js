import { 
    auth,
    signInWithPopup,
    googleAuthProvider,
    createUserWithEmailAndPassword,
    updateProfile,
    signInWithEmailAndPassword,
    signOut
} from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoader, startLoader } from "./loaderActions";
import Swal from 'sweetalert2';


export const startLoginEmailPassword = (email,password) =>{
    return async (dispatch)=>{

        await dispatch(startLoader());
        
        await signInWithEmailAndPassword(auth,email,password)
        .then((userCred) =>{

            const  {user:{uid,displayName,photoURL}} = userCred;
            dispatch(login(uid,displayName,photoURL));

        })
        .catch(err =>{
            console.log(err);
            Swal.fire({
                title: 'Error!',
                text: err,
                icon: 'error',
                confirmButtonText: 'ok'
            })
        })

        await dispatch(finishLoader());
    }
};

export const startGoogleLogin = () =>{
    return (dispatch)=>{
        signInWithPopup(auth,googleAuthProvider)
        .then( userCred =>{
            //console.log(userCred);
            const  {user:{uid,displayName,photoURL}} = userCred;
            dispatch(login(uid,displayName,photoURL));
        })
    }
};

export const login = (uid,displayName,photoURL='https://placeimg.com/100/100/animals') => (
    {
        type: types.login,
        payload:{
            uid,
            displayName,
            photoURL
        }
    }
);

export const registerUser = (email,password,userName,reset)=>{
    return async (dispatch)=>{
        
        await dispatch(startLoader());

        //crear usuario en firebase
        await createUserWithEmailAndPassword(auth,email,password)
        .then( async ({user}) => {
            
            //funcion defirebase al obtener el usuario
            //permite registrar el nombre y la foto
            await updateProfile(auth.currentUser,{
                displayName: userName,
                photoURL: 'https://placeimg.com/100/100/any'
            });

            //actualizar el estado global
            const  {uid,displayName,photoURL} = user;
            dispatch(login(uid,displayName,photoURL));

            //reset form
            reset();

        })
        .catch(err =>{
            Swal.fire({
                title: 'Error!',
                text: err,
                icon: 'error',
                confirmButtonText: 'ok'
            })
        })

        await dispatch(finishLoader());
        
    }
};

export const startLogout = ()=>{
    return async (dispatch) =>{
        await signOut(auth);
        dispatch(logout());
    }
}

export const logout = ()=>(
    {
        type: types.logout
    }
)
