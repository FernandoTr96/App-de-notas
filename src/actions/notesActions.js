import Swal from "sweetalert2";
import { db, addDoc, collection, getDocs, updateDoc, doc, deleteDoc, query, orderBy } from "../firebase/firebase-config";
import { uploadImg } from "../helpers/uploadImg";
import {removeImg} from '../helpers/removeImg'
import { types } from "../types/types";
import { finishLoader, startLoader } from "./loaderActions";

//insertar en firestore addDoc

/* PARA INSERTAR EN FIRESTORE DEBE ESTAR AUTENTICADO    
    allow read, write: if request.auth != null;
*/

export const startNewNote = () => {
    //getState: permite traer el estado global como un useSelector()
    //estos argumentos son pasados por el moddleware
    return async (dispatch, getState) => {

        //id del usuario que llamo la accion o esta en sesion
        const uid = getState().auth.uid;

        //objeto que contendra la info de la nota
        const newNote = {
            title: 'New note',
            body: '...',
            date: new Date().getTime()
        }

        //insertar en firestore
        try {

            dispatch(startLoader());

            const docRef = await addDoc(
                collection(db, `${uid}/journal/notes`),
                newNote
            );

            //activar nota
            dispatch(activeNote(docRef.id, newNote));

            //obtener las notas para actualizar el estado
            dispatch(getNotes(uid));

            dispatch(finishLoader());


        } catch (error) {

            console.log(error);
            Swal.fire({
                title: 'Not created',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }

    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const resetActiveNote = () => ({
    type: types.notesResetActive
})


//Leer datos de firestore getDocs

export const getNotes = (uid) => {
    return async (dispatch) => {

        let notes = [];

        try {

            const q = query(collection(db, `${uid}/journal/notes`), orderBy('date', 'desc'));
            const data = await getDocs(q);

            data.forEach((doc) => {
                notes.push(
                    {
                        id: doc.id,
                        ...doc.data()
                    }
                )
            })

            dispatch(setNotes(notes));

        } catch (error) {

            Swal.fire({
                title: 'I cant get the notes',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }

    }
}

export const setNotes = (notes) => (
    {
        type: types.notesSet,
        payload: notes
    }
);


//actualizar nota
export const updateNote = (note) => {
    return async (dispatch, getState) => {

        const { auth: { uid } } = getState();
        const noteToFireStore = { ...note };
        delete noteToFireStore.id;

        try {

            const docRef = doc(db, `${uid}/journal/notes`, note.id);
            await updateDoc(docRef, noteToFireStore);
            dispatch(activeNote(note.id, note));
            dispatch(refreshNote(note.id, note));

            Swal.fire({
                title: 'Note updated',
                text: 'The note has been updated successfully !!',
                icon: 'success',
                confirmButtonText: 'ok'
            })

        } catch (error) {

            Swal.fire({
                title: 'I cant update the note',
                text: error,
                icon: 'error',
                confirmButtonText: 'ok'
            })

        }
    }
}

const refreshNote = (id, data) => ({
    type: types.noteRefresh,
    payload: {
        id, data
    }
})


//subir imagen
export const uploadImage = (file) => {
    return async (dispatch, getState) => {

        dispatch(startLoader());

        const upload = await uploadImg(file);

        if (upload !== null) {

            const {url, public_id} = upload;

            const noteActive = getState().notes.active;
            const noteWithUrlToFirestore = {
                ...noteActive,
                public_id,
                url
            };

            if (getState().notes.active?.public_id) {
                const public_id = getState().notes.active.public_id;
                removeImg(public_id);
            }

            dispatch(activeNote(noteWithUrlToFirestore.id, noteWithUrlToFirestore));
            dispatch(updateNote(noteWithUrlToFirestore));

        }

        dispatch(finishLoader());

    }
}

//Eliminar nota

export const deleteNote = (id) => {
    return (dispatch, getState) => {

        Swal.fire({
            title: 'Are you sure?',
            text: "You won't be able to revert this!",
            icon: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes, delete it!'
        }).then(async (result) => {
            if (result.isConfirmed) {

                dispatch(startLoader());

                try {

                    const uid = getState().auth.uid;
                    await deleteDoc(doc(db, `${uid}/journal/notes`, id));
                    dispatch(removeNote(id));

                    if (getState().notes.active?.public_id) {
                        const public_id = getState().notes.active.public_id;
                        removeImg(public_id);
                    }

                    Swal.fire(
                        'Deleted!',
                        'Your note has been deleted.',
                        'success'
                    )


                } catch (error) {

                    Swal.fire({
                        title: 'i cant delete the note',
                        text: error,
                        icon: 'error',
                        confirmButtonText: 'ok'
                    })

                }

                dispatch(finishLoader());

                dispatch(resetActiveNote());

            }
        })
    }
}

const removeNote = (id) => (
    {
        type: types.noteRemove,
        payload: id
    }
)

export const purgeNotes = () => (
    {
        type: types.purgeNotes
    }
)

/* ASI ENTIENDO NO SQL

//coleccion
uid = [
    //documento
    journal = {

        //coleccion
        notes = [

            //documento
            id = {

                //campos
                body: "",
                title: "",
                date: ""
            }
        ]
    }
]

[
    {
        journal: {
            notes:[
                01:{
                    body: "",
                    title: "",
                    date: ""
                }
            ]
        }
    }
]*/