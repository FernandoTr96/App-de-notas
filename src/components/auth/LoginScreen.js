import React from 'react';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/authActions';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from "react-hook-form";
import { Loader } from '../loading/Loader';

/* validar formularios con el paquete : react-hooks-form
   https://react-hook-form.com/get-started
*/

// para instalar el manejador de firebase en mi app y poder hacer autenticacion
// npm install firebase


export const LoginScreen = () => {

    //paquete de validaciones
    const { register, handleSubmit, watch, getValues, formState: { errors } } = useForm();

    //console.log ( useForm() );

    //para hacer dispatch a mi reducer de la store se utiliza
    //el hook useDispatch de react-redux
    const dispatch = useDispatch();

    // useSelector de redux permite obtener el estado global
    // osea el conjunto de reducers
    const { loader: { loading } } = useSelector(state => state);

    const submit = () => {
        //valores del formulario
        const { email, password } = getValues();

        //dispatcch de la accion login
        dispatch(startLoginEmailPassword(email, password));
    };

    const handleGoogleSingIn = () => {
        dispatch(startGoogleLogin());
    }

    return (

        <form onSubmit={handleSubmit(submit)} className='auth__form animate__animated animate__fadeIn' autoComplete='off'>

            <legend>
                JournalApp
            </legend>

            <input
                type='text'
                placeholder='ðŸ“§ example@gmail.com'
                defaultValue={watch('email')}
                {...register('email', { required: true, pattern: /^[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?\.)+[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?$/i })}
            />
            {errors.email?.type === 'required' && <small className='error'>*This field is required</small>}
            {errors.email?.type === 'pattern' && <small className='error'>*The email does not have a valid structure</small>}

            <input
                type='password'
                placeholder='ðŸ”‘ password'
                defaultValue={watch('password')}
                {...register('password', { required: true })}
            />
            {errors.password && <small className='error'>*This field is required</small>}

            <button type='submit' disabled={loading} >
                {loading && <Loader src='https://samherbert.net/svg-loaders/svg-loaders/tail-spin.svg' />}
                Log-in
            </button>

            <div onClick={handleGoogleSingIn} className='auth__google-sign-in'>
                <p className='auth__google-legend'>Login with social network</p>
                <div className='auth__google-button'>
                    <div>
                        <img
                            src='https://freesvg.org/img/1534129544.png'
                            alt='google'
                        />
                    </div>
                    <p>
                        <b>Sign in with google</b>
                    </p>
                </div>
            </div>

            <Link to='/auth/register' className='auth__link'>
                Create new account
            </Link>

        </form>

    )
}
