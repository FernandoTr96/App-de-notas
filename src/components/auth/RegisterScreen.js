import React from 'react'
import {Link} from 'react-router-dom'
import {useForm} from "react-hook-form"
import {useDispatch, useSelector} from 'react-redux'
import { registerUser } from '../../actions/authActions';
import { Loader } from '../loading/Loader';


/* validar formularios con el paquete : react-hooks-form
   https://react-hook-form.com/get-started
*/

export const RegisterScreen = () => {

    //paquete de validaciones
    const {register, handleSubmit, reset, watch, getValues, formState: { errors }} = useForm();

    //console.log ( useForm() );

    const dispatch = useDispatch();

    const {loader:{loading}} = useSelector(state => state);

    const submit = ()=>{
        const {email,password,userName} = getValues();
        dispatch(registerUser(email,password,userName,reset));
    };
    

    return (
        <form onSubmit={handleSubmit(submit)} className='auth__form animate__animated animate__fadeIn' autoComplete='off'>

            <legend>
                Register
            </legend>

            <input
                type='email'
                placeholder='üìß example@gmail.com'
                defaultValue={watch('email')}
                { ...register( 'email' , {required:true, pattern: /^[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[A-Za-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?\.)+[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?$/i} ) }
            />
            {errors.email?.type==='required' && <small className='error'>*Este campo es requerido</small>}
            {errors.email?.type==='pattern' && <small className='error'>*La estructura del correo no es valida</small>}

            <input
                type='text'
                placeholder='üë§ your user name'
                defaultValue={watch('userName')}
                { ...register( 'userName' , {required:true} ) }
            />
            {errors.userName?.type==='required' && <small className='error'>*Este campo es requerido</small>}

            <input
                type='password'
                placeholder='üîë Password'
                defaultValue={watch('password')}
                { ...register( 'password' , {required:true, pattern: /^(?=\w*\d)(?=\w*[A-Z])(?=\w*[a-z])\S{8,16}$/i} ) }
            />
            {errors.password?.type==='required' && <small className='error'>*Este campo es requerido</small>}
            {errors.password?.type==='pattern' && <small className='error'>*La contrase√±a debe tener al entre 8 y 16 caracteres, al menos un d√≠gito, al menos una min√∫scula y al menos una may√∫scula</small>}

            <input
                type='password'
                placeholder='üîë Confirm password'
                defaultValue={watch('confirmPwd')}
                { ...register( 'confirmPwd' , {required:true} ) }
            />
            {errors.confirmPwd?.type==='required' && <small className='error'>*Este campo es requerido</small>}
            {watch('password') !== watch('confirmPwd') && <small className='error'>*Las contrase√±as no coinciden</small>}
            
            <button type='submit' disabled={loading}>
            {loading && <Loader src = 'https://samherbert.net/svg-loaders/svg-loaders/tail-spin.svg' /> } 
                Create account
            </button>

            <Link to='/auth/login' className='auth__link'>
                I have a account 
            </Link>

        </form>
    )
}
