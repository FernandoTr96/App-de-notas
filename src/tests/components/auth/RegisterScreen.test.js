import React from 'react';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import {mount} from 'enzyme'
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import {act} from 'react-dom/test-utils'
import { registerUser } from '../../../actions/authActions';

const middlewares = [thunk];
const mockstore = configureStore(middlewares);
const initState = {
    loader: {
        loading: false
    }
};
const store = mockstore(initState);

//Actions must be plain objects. Use custom middleware for async actions. 
//saltamos el dispatch y lo tomaos como una funcion x
store.dispatch = jest.fn();

let wrapper = mount(
    <Provider store={store} >
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);

//mock actions
jest.mock('../../../actions/authActions', () => ({
    registerUser: jest.fn()
}));



describe('Pruebas en <RegisterScreen/>', () => {

    beforeEach(()=>{
        wrapper = mount(
            <Provider store={store} >
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );
    })

    test('debe pintarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    });
    
    test('el submit debe disparar la accion de registro de ususario', async () => {
        
        await act( async ()=> {

            wrapper.find('input').at(0).simulate('change',{
                target: {
                    name: 'email',
                    value: 'test@gmail.com'
                }
            })

            wrapper.find('input').at(1).simulate('change',{
                target: {
                    name: 'userName',
                    value: 'Fernando96'
                }
            })

            wrapper.find('input').at(2).simulate('change',{
                target: {
                    name: 'password',
                    value: '54321haloO'
                }
            })

            wrapper.find('input').at(3).simulate('change',{
                target: {
                    name: 'confirmPwd',
                    value: '54321haloO'
                }
            })

            wrapper.find('.auth__form').simulate('submit');

        })

        expect(registerUser).toHaveBeenCalled();

    });


    test('los mensajes de error deben existir', async () => {
        
        await act( async ()=> {

            wrapper.find('input').at(0).simulate('change',{
                target: {
                    name: 'email',
                    value: 'emailerroneo.com'
                }
            })

            wrapper.find('input').at(2).simulate('change',{
                target: {
                    name: 'password',
                    value: '123erroneo'
                }
            })

            wrapper.find('.auth__form').simulate('submit');

        })
   
        expect(wrapper.html().includes('error')).toBe(true);

    });


    test('el link debe redireccionar a el login ', () => {
        
        const href = wrapper.find('Link').prop('to');
        expect(href).toBe('/auth/login')

    });

});