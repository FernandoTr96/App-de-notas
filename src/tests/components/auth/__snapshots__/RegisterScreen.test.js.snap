// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen/> debe pintarse correctamente 1`] = `
<form
  autoComplete="off"
  className="auth__form animate__animated animate__fadeIn"
  onSubmit={[Function]}
>
  <legend>
    Register
  </legend>
  <input
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="ðŸ“§ example@gmail.com"
    type="email"
  />
  <input
    name="userName"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="ðŸ‘¤ your user name"
    type="text"
  />
  <input
    name="password"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="ðŸ”‘ Password"
    type="password"
  />
  <input
    name="confirmPwd"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="ðŸ”‘ Confirm password"
    type="password"
  />
  <button
    disabled={false}
    type="submit"
  >
    Create account
  </button>
  <a
    className="auth__link"
    href="/auth/login"
    onClick={[Function]}
  >
    I have a account
  </a>
</form>
`;
